lib/spack/spack/spec.py

```diff
@@ -4413,7 +4413,7 @@ def __hash__(self):
         return hash(lang.tuplify(self._cmp_iter))
 
     def __reduce__(self):
-        return _spec_from_dict, (self.to_dict(hash=ht.build_hash),)
```

Comment:
I think this may be wrong, at least in `develop`, in the sense that using the default means using the DAG hash. The information returned here should be enough to reconstruct an equivalent `Spec` object when we unpickle (and I think `pickle` caches these results, though I can look for a definitive confirmation of that). 

Since we serialize to pass the information to a spawned process on macOS and we don't seek persistence, I felt like a change in the recipe was ruled out and build hash was the right granularity. Using the dag hash will make any spec with different build deps equivalent to each other.

Reply:
I think using the build hash makes sense. However, this PR subtly changes `to_dict` so that when the caller specifies a hash, it writes and "indexes by" (checks node uniqueness by) that hash. I was trying to avoid an implementation that looks at all the nodes and sees if they all have a full hash, then a build hash, then just go by DAG hash, but that might be necessary.